src/toolchain.ml

```diff
@@ -135,59 +127,71 @@ let get_install_command { opam = opam, switch; _ } tools =
   let opam_packages = List.map tools ~f:Tool.opam_package in
   Opam.install opam switch ~packages:opam_packages
 
-let install_missing_tools t tools opamroot =
+let with_progress ~msg f =
   let open Promise.Syntax in
-  match tools with
-  | [] -> Promise.return ()
-  | tools ->
-    let progress_options =
-      ProgressOptions.create ~location:(`ProgressLocation Window)
-        ~title:"Installing Platform Tools (this might take a few minutes)"
-        ~cancellable:false ()
-    in
-    let task ~progress:_ ~token:_ =
-      let* is_installed = is_sandbox_installed t in
-      let* _ =
-        match is_installed with
-        | false -> setup_toolchain_sandbox t.project_sandbox opamroot
-        | true -> Promise.Result.return ()
-      in
-      let cmd = get_install_command t tools in
-      let+ result = Cmd.output cmd in
-      match result with
-      | Ok _ ->
-        show_message `Info "The platform tools have been successfully installed";
-        Ojs.null
-      | Error err ->
-        show_message `Error "The installation of the Platform tools failed: %s"
-          err;
-        Ojs.null
-    in
-    let+ (_ : Ojs.t) =
-      Window.withProgress (module Ojs) ~options:progress_options ~task
-    in
-    ()
-
-let upgrade t =
-  let open Promise.Syntax in
-  let opam, switch = t.opam in
   let progress_options =
-    ProgressOptions.create ~location:(`ProgressLocation Window)
-      ~title:"Checking for Platform Tools updates" ~cancellable:false ()
+    ProgressOptions.create ~location:(`ProgressLocation Window) ~title:msg
+      ~cancellable:false ()
   in
   let task ~progress:_ ~token:_ =
-    let* _ = Opam.update opam switch |> Cmd.output in
-    let+ result = Opam.upgrade opam switch |> Cmd.output in
+    let+ result = f () in
     match result with
-    | Ok _ -> Ojs.null
+    | Ok _ -> Ok ()
     | Error err ->
-      show_message `Error "The upgrade of the Platform tools failed: %s" err;
-      Ojs.null
-  in
-  let+ (_ : Ojs.t) =
-    Window.withProgress (module Ojs) ~options:progress_options ~task
+      show_message `Error "%s" err;
+      Error err
   in
-  ()
+  Window.withProgress
+    (module Interop.Js.Result (Interop.Js.Unit))
+    ~options:progress_options ~task
+
+let opam_init opam opamroot =
+  let open Promise.Syntax in
+  let* exists = Fs.exists (opamroot |> Path.to_string) in
+  if exists then
+    Promise.Result.return ()
+  else
+    with_progress
+      ~msg:"Installing Platform Tools (this might take a few minutes)"
```

Comment:
I think this should have a different message from `install_missing_tools` so the user has a little more information.

Maybe `Initializing Platform Toolchain (this might take a few minutes)`?